src/cabin/cabin.service.ts

```diff
@@ -1,14 +1,115 @@
+  private readonly visbookApiClient = new VisbookApi();
+
   async getRandomCabin(): Promise<CabinDetails> {
     const cabins = await this.apiClient.getCabins();
     const cabin = cabins[Math.floor(Math.random() * cabins.length)];
     const { id } = cabin.node;
     const cabinDetails = await this.apiClient.getCabinDetails(id);
     return cabinDetails;
   }
+
+  async isCabinAvailable(
+    cabin: CabinDetails,
+    checkIn: string,
+    checkOut: string,
+  ): Promise<boolean> {
+    if (!dateIsValid(checkIn) || !dateIsValid(checkOut)) {
+      console.log('invalid date(s)');
+      return false;
+    }
+
+    if (cabin.bookingUrl === null) {
+      console.log('no booking url');
+      return false;
+    }
+
+    if (!cabin.bookingUrl.includes('reservations')) {
+      console.log('not a reservations url');
+      return false;
+    }
+
+    const cabinVisbookId = getVisbookId(cabin.bookingUrl);
+
+    if (cabinVisbookId === 0) return false;
+
+    const cabinIsAvailable = await this.visbookApiClient.isCabinAvailable(
+      cabinVisbookId,
+      checkIn,
+      checkOut,
+    );
+
+    return cabinIsAvailable;
+  }
+
+  async getRandomAvailableCabin(
+    checkIn: string,
+    checkOut: string,
+  ): Promise<CabinDetails | undefined> {
+    if (!dateIsValid(checkIn) || !dateIsValid(checkOut)) {
+      return undefined;
+    }
+
+    const hasExceededTimeLimit = (now: number, timeLimit: number): boolean => {
+      const elapsedTime = now - startTime;
+      if (elapsedTime >= timeLimit) {
+        return true;
+      }
+      return false;
+    };
+
+    const startTime = Date.now();
+    let cabin = await this.getRandomCabin();
+
+    while ((await this.isCabinAvailable(cabin, checkIn, checkOut)) === false) {
+      cabin = await this.getRandomCabin();
+      if (hasExceededTimeLimit(Date.now(), 60000)) {
+        /* TODO
+        - How long should response time (time limit) be? 1min? 5min?
+        - Do we have a limit on how many api calls we can make in a given time?
+        - Return undefined or specific message aka "could not find within time limit?"
+        */
+        return undefined;
+      }
+    }
+
+    return cabin;
+  }
+
+  async getAvailableCabins(
+    checkIn: string,
+    checkOut: string,
+  ): Promise<CabinDetails[] | undefined> {
+    if (!dateIsValid(checkIn) || !dateIsValid(checkOut)) {
+      return undefined;
+    }
+
+    const cabins = await this.apiClient.getCabins();
+
+    // TODO we need cabin details here or change where cabin details are fetched
+    /*const areCabinsAvailable = await Promise.all(
+      cabins.map((cabin) =>
+        this.isCabinAvailable(cabin.node.id, checkIn, checkOut),
+      ),
+    );
+
+    const availableCabins = cabins.filter(
+      (_value, index) => areCabinsAvailable[index],
+    );
+
+    const cabinDetails = await Promise.all(
+      availableCabins.map((cabin) =>
+        this.apiClient.getCabinDetails(cabin.node.id),
+      ),
+    );
+
+    return cabinDetails;*/
```

Comment:
I am not a big fan of sleeping code as requirements usually change pretty quickly and this sort of thing becomes less valuable.